solana/solana-ibc/programs/solana-ibc/src/lib.rs

```diff
@@ -191,6 +201,68 @@ pub mod solana_ibc {
         chain.set_stake((validator).into(), amount)
     }
 
+    /// Sets up new fee collector proposal which wont be changed until the new fee collector
+    /// calls `accept_fee_collector_change`. If the method is called for the first time, the fee
+    /// collector would just be set without needing for any approval.
+    ///
+    pub fn setup_fee_collector<'a, 'info>(
+        ctx: Context<'a, 'a, 'a, 'info, SetupFeeCollector<'info>>,
+        new_fee_collector: Pubkey,
+    ) -> Result<()> {
+        let private_storage = &mut ctx.accounts.storage;
+
+        let signer = ctx.accounts.fee_collector.key();
+
+        if private_storage.fee_collector == Pubkey::default() {
+            private_storage.fee_collector = new_fee_collector;
+        } else if signer == private_storage.fee_collector {
+            private_storage.new_fee_collector_proposal = Some(new_fee_collector)
+        } else {
+            return Err(error::Error::InvalidFeeCollector)
+                .map_err(|err| error!((&err)));
+        }
+
+        Ok(())
+    }
+
+    pub fn accept_fee_collector_change<'a, 'info>(
+        ctx: Context<'a, 'a, 'a, 'info, SetupFeeCollector<'info>>,
+    ) -> Result<()> {
+        let private_storage = &mut ctx.accounts.storage;
+
+        let signer = ctx.accounts.fee_collector.key();
+
+        if let Some(new_admin) = private_storage.new_fee_collector_proposal {
+            if signer != new_admin {
+                return Err(error::Error::InvalidFeeCollector)
+                    .map_err(|err| error!((&err)));
+            }
+            private_storage.fee_collector = new_admin;
+            private_storage.new_fee_collector_proposal = None;
+        } else {
+            return Err(error::Error::FeeCollectorChangeProposalNotSet)
+                .map_err(|err| error!((&err)));
+        }
+
+        Ok(())
+    }
```

Comment:
I think we should just require a transaction signed by both. It makes things easier on the contract side. And on the client side it would effectively work as multisig. Does that make sense?

Reply:
but its 2 different accounts signing on those methods. The first one is signed by existing one and the second is signed by the new fee collector.

Reply:
Yes, but transaction can be signed by multiple accounts so it should be possible. The difference is that the accounts holders would need to coordinate off-chain to execute the transaction.

Reply:
i think it would be a bit convoluted though

Reply:
for the users.

Reply:
Do we change fee collector often? I’d rather move complexity to client than having it in the contract. Solana has interface for partial signing: https://solana-labs.github.io/solana-web3.js/classes/Transaction.html#partialSign.partialSign-1

Reply:
well ideally not

Reply:
i think partial sign could work

Reply:
just worried how we would do with multisig

Reply:
closing this as we decided that it can get complicated